src/hotspot/share/nmt/vmatree.cpp

```diff
@@ -0,0 +1,242 @@
+      // state change, we just omit the node.
+      // That happens, for example, when reserving within an already reserved region with identical metadata.
+      stA.in = leqA_n->val().out; // .. and the region's prior state is the incoming state
+      if (stA.is_noop()) {
+        // Nothing to do.
+      } else {
+        // Add new node.
+        tree.upsert(A, stA);
+      }
+    }
+  }
+
+  // Now we handle B.
+  // We first search all nodes that are (A, B]. All of these nodes
+  // need to be deleted and summary accounted for. The last node before B determines B's outgoing state.
+  // If there is no node between A and B, its A's incoming state.
+  GrowableArrayCHeap<AddressState, mtNMT> to_be_deleted_inbetween_a_b;
+  bool B_needs_insert = true;
+
+  // Find all nodes between (A, B] and record their addresses. Also update B's
+  // outgoing state.
+  { // Iterate over each node which is larger than A
+    GrowableArrayCHeap<VTreap*, mtNMT> to_visit;
+    to_visit.push(tree._root);
+    VTreap* head = nullptr;
+    while (!to_visit.is_empty()) {
+      head = to_visit.pop();
+      if (head == nullptr) continue;
+
+      int cmp_A = AddressComparator::cmp(head->key(), A);
+      int cmp_B = AddressComparator::cmp(head->key(), B);
+      if (cmp_B > 0) {
+        // head > B
+        to_visit.push(head->left());
+      } else if (cmp_A <= 0) {
+        // head <= A
+        to_visit.push(head->right());
+      } else if (cmp_A > 0 && cmp_B <= 0) {
+        // A < head <= B
+        to_visit.push(head->left());
+        to_visit.push(head->right());
+
+        stB.out = head->val().out;
+        if (cmp_B < 0) {
+          // Record all nodes preceding B.
+          to_be_deleted_inbetween_a_b.push({head->key(), head->val()});
+        } else if (cmp_B == 0) {
+          // Re-purpose B node, unless it would result in a noop node, in
+          // which case record old node at B for deletion and summary accounting.
+          if (stB.is_noop()) {
+            to_be_deleted_inbetween_a_b.push(AddressState{B, head->val()});
+          } else {
+            head->val() = stB;
+          }
+          B_needs_insert = false;
+        } else {
+          assert(false, "Cannot happen.");
+        }
+      } else {
+        // Impossible.
+        assert(false, "Cannot happen.");
+      }
+    }
+  }
+
+  // Insert B node if needed
+  if (B_needs_insert && // Was not already inserted
+      !stB.is_noop())   // The operation is differing
+    {
+    tree.upsert(B, stB);
+  }
+
+  // Finally, we need to:
+  // 1. Perform summary accounting.
+  // 2. Delete all nodes between (A, B]. Including B in the case of a noop.
+
+  if (to_be_deleted_inbetween_a_b.length() == 0 && LEQ_A_found && GEQ_B_found) {
+    // We have smashed a hole in an existing region (or replaced it entirely).
+    // LEQ_A - A - B - GEQ_B
+    auto& rescom = diff.flag[NMTUtil::flag_to_index(LEQ_A.flag_out())];
+    if (LEQ_A.state.out.type() == StateType::Reserved) {
```

Comment:
I prefer showing the explicit comparison being done, instead of hiding it under a function or method. Just to make it obvious that there's no extra logic being performed. I did shorten the code by adding `out()` and `in()` accessors.